### PR TITLE
[Merged by Bors] - feat: port Algebra.Module.Injective

### PR DESCRIPTION
**Pull request by mattrobball**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Injective.lean [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+instance : SemilatticeInf (ExtensionOf i f) :=
+  Function.Injective.semilatticeInf ExtensionOf.toLinearPMap
+    (fun X Y h =>
+      ExtensionOf.ext (by rw [h]) fun x y h' => by
+        have : {x y : N} → (h'' : x = y) → (hx : x ∈ X.toLinearPMap.domain) →
+          (hy : y ∈ Y.toLinearPMap.domain) → X.toLinearPMap ⟨x,hx⟩ = Y.toLinearPMap ⟨y,hy⟩ := by
+            rw [h]
+            intro _ _ h _ _
+            congr
+        apply this h' _ _)
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:162]: Maybe add a porting note about the changes you made to the Lean 3 proof? I thought the philosophy was that if you deviate from the script, then leave a porting note.

File: Mathlib/Algebra/Module/Injective.lean [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+variable {f}
+
+private theorem extensionOfMax_adjoin.aux1 {y : N} (x : supExtensionOfMaxSingleton i f y) :
+    ∃ (a : (extensionOfMax i f).domain) (b : R), x.1 = a.1 + b • y := by
+  have mem1 : x.1 ∈ (_ : Set _) := x.2
+  rw [Submodule.coe_sup] at mem1
+  rcases mem1 with ⟨a, b, a_mem, b_mem : b ∈ (Submodule.span R _ : Submodule R N), eq1⟩
+  rw [Submodule.mem_span_singleton] at b_mem
+  rcases b_mem with ⟨z, eq2⟩
+  exact ⟨⟨a, a_mem⟩, z, by rw [← eq1, ← eq2]⟩
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:265]: Is the idea that we don't need to #align private theorems?

File: Mathlib/Algebra/Module/Injective.lean [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+theorem ExtensionOfMaxAdjoin.eqn {y : N} (x : supExtensionOfMaxSingleton i f y) :
+    ↑x = ↑(ExtensionOfMaxAdjoin.fst i x) + ExtensionOfMaxAdjoin.snd i x • y :=
+  (extensionOfMax_adjoin.aux1 i x).choose_spec.choose_spec
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_adjoin.eqn Module.Baer.ExtensionOfMaxAdjoin.eqn
+
+variable (f)
+
+set_option synthInstance.etaExperiment true in
+-- Porting note: we don't have colon ideals!?
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:289]:
```lean
/-- `N.colon P` is the ideal of all elements `r : R` such that `r • P ⊆ N`. -/
def colon (N P : Submodule R M) : Ideal R :=
  annihilator (P.map N.mkQ)
#align submodule.colon Submodule.colon
```

In RingTheory.Ideal.Operations (which is ported)

File: Mathlib/Algebra/Module/Injective.lean [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_adjoin.extend_ideal_to_wd Module.Baer.ExtensionOfMaxAdjoin.extendIdealTo_wd
+
+set_option synthInstance.etaExperiment true in
+theorem ExtensionOfMaxAdjoin.extendIdealTo_eq (h : Module.Baer R Q) {y : N} (r : R)
+    (hr : r • y ∈ (extensionOfMax i f).domain) : ExtensionOfMaxAdjoin.extendIdealTo i f h y r =
+    (extensionOfMax i f).toLinearPMap ⟨r • y, hr⟩ := by
+  simp only [ExtensionOfMaxAdjoin.extendIdealTo_is_extension i f h _ _ hr,
+    ExtensionOfMaxAdjoin.idealTo, LinearMap.coe_mk, Subtype.coe_mk]
+  dsimp only [AddHom.coe_mk]
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:350]:
```suggestion
  -- porting note: in mathlib3 `AddHom.coe_mk` was not needed
  simp only [ExtensionOfMaxAdjoin.extendIdealTo_is_extension i f h _ _ hr,
    ExtensionOfMaxAdjoin.idealTo, LinearMap.coe_mk, Subtype.coe_mk, AddHom.coe_mk]
```

File: Mathlib/Algebra/Module/Injective.lean [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+          rw [ExtensionOfMaxAdjoin.eqn, ExtensionOfMaxAdjoin.eqn, add_smul, Submodule.coe_add]
+          ac_rfl
+        rw [ExtensionOfMaxAdjoin.extensionToFun_wd (y := y) i f h (a + b) _ _ eq1,
+          LinearPMap.map_add, map_add]
+        unfold ExtensionOfMaxAdjoin.extensionToFun
+        abel
+      map_smul' := fun r a => by
+        dsimp
+        have eq1 :
+          r • (a:N) =
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:411]:
```suggestion
          r • (a : N) =
```

File: Mathlib/Algebra/Module/Injective.lean [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+            ↑(r • ExtensionOfMaxAdjoin.fst i a) + (r • ExtensionOfMaxAdjoin.snd i a) • y := by
+          rw [ExtensionOfMaxAdjoin.eqn, smul_add, smul_eq_mul, mul_smul]
+          rfl
+        rw [ExtensionOfMaxAdjoin.extensionToFun_wd i f h (r • a) _ _ eq1, LinearMap.map_smul,
+          LinearPMap.map_smul, ← smul_add]
+        congr }
+  is_extension m := by
+    dsimp
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:419]: Do you understand why this is needed in mathlib4 but not mathlib3? Is it worth a porting note?

File: Mathlib/Algebra/Module/Injective.lean [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+    change ExtensionOfMaxAdjoin.extensionToFun i f h _ = _
+    rw [ExtensionOfMaxAdjoin.extensionToFun_wd i f h x' x 0 (by simp [EQ]), map_zero,
+      add_zero]⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_le Module.Baer.extensionOfMax_le
+
+theorem extensionOfMax_to_submodule_eq_top (h : Module.Baer R Q) :
+    (extensionOfMax i f).domain = ⊤ := by
+  refine' Submodule.eq_top_iff'.mpr fun y => _
+  dsimp
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:441]: Similar comment here. Is the fact that we need `dsimp` evidence of some bigger problem somehow?

File: Mathlib/Algebra/Module/Injective.lean [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+  rw [← extensionOfMax_is_max i f _ (extensionOfMax_le i f h), extensionOfMaxAdjoin]
+  rw [Submodule.mem_sup]
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:443]:
```suggestion
  rw [← extensionOfMax_is_max i f _ (extensionOfMax_le i f h), extensionOfMaxAdjoin,
    Submodule.mem_sup]
```

File: Mathlib/Algebra/Module/Injective.lean [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+      haveI : Fact (Function.Injective i) := ⟨hi⟩
+      ⟨{  toFun := fun y =>
+            (extensionOfMax i f).toLinearPMap
+              ⟨y, (extensionOfMax_to_submodule_eq_top i f h).symm ▸ trivial⟩
+          map_add' := fun x y => by
+            rw [← LinearPMap.map_add]
+            congr
+          map_smul' := fun r x => by
+            rw [← LinearPMap.map_smul]
+            dsimp },
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/Injective.lean:461]:
```suggestion
            -- porting note: was `congr`
            dsimp },
```
I also had a `congr` fail on me today in a ported file. Is it worth keeping track of these failures?

**Review by kbuzzard — COMMENTED** [review]
This looks fine. I raise many pernickety points, which can probably be ignored.

File: Mathlib/Algebra/Module/Injective.lean [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+* `Module.Baer.injective`: an `R`-module is injective if it is Baer.
+
+-/
+
+
+noncomputable section
+
+universe u v
+
+variable (R : Type u) [Ring R] (Q : TypeMax.{v,u}) [AddCommGroup Q] [Module R Q]
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:46]: I don't think the order is important here for behavior (but should be checked on this). What is the newly established convention? I still had the category theory ordering in my mind.

File: Mathlib/Algebra/Module/Injective.lean [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+
+variable {R Q} {M N : Type max u v} [AddCommGroup M] [AddCommGroup N]
+
+variable [Module R M] [Module R N] (i : M →ₗ[R] N) (f : M →ₗ[R] Q)
+
+/-- If we view `M` as a submodule of `N` via the injective linear map `i : M ↪ N`, then a submodule
+between `M` and `N` is a submodule `N'` of `N`. To prove Baer's criterion, we need to consider
+pairs of `(N', f')` such that `M ≤ N' ≤ N` and `f'` extends `f`. -/
+structure ExtensionOf extends LinearPMap R N Q where
+  le : LinearMap.range i ≤ domain
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:110]: Yeah `i.range` was fine in Lean 3. I've seen a few failures of dot notation in the `LinearMap` namespace. I didn't really investigate but just powered on.

File: Mathlib/Algebra/Module/Injective.lean (hunk 6 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:419]: I think is the change in default transparency.

File: Mathlib/Algebra/Module/Injective.lean (hunk 7 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:441]: I think it is just the world we live in now.

File: Mathlib/Algebra/Module/Injective.lean (hunk 9 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:461]: Probably. I'll record something.

File: Mathlib/Algebra/Module/Injective.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:162]: I will once I remember why...

**Comment by mattrobball** [comment]
I'll push changes once GitHub stands back up completely

File: Mathlib/Algebra/Module/Injective.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:162]: Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 4 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:350]: Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 5 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:411]: Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 8 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:443]: Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 9 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:461]: Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:265]: Fixed

File: Mathlib/Algebra/Module/Injective.lean (hunk 11 above)
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Module/Injective.lean:110]: In a previous case, all it took was reordering some arguments in a `variable` declaration where the similarly misbehaving declaration to `LinearMap.range` lived.

**Comment by mattrobball** [comment]
> I'll push changes once GitHub stands back up completely

Done

File: Mathlib/Algebra/Module/Injective.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/Algebra/Module/Injective.lean:293]: So I think this porting note should be updated...

**Review by jcommelin — COMMENTED** [review]
Please update the porting note about colon ideals. Either remove it, or make it a `Porting note: TODO refactor using colon ideals` or something like that.

bors d+

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**